kratos/geometries/quadrilateral_3d_4.h

```diff
@@ -925,16 +909,18 @@ template<class TPointType> class Quadrilateral3D4
      * @see Jacobian
      * @see InverseOfJacobian
      */
-    double DeterminantOfJacobian( IndexType IntegrationPointIndex,
-                                          IntegrationMethod ThisMethod ) const override
+    double DeterminantOfJacobian( 
+        IndexType IntegrationPointIndex,
+        IntegrationMethod ThisMethod 
+        ) const override
     {
         Matrix jacobian ( 3, 2 );
          
-        this->Jacobian( jacobian, IntegrationPointIndex, ThisMethod);
+        Jacobian( jacobian, IntegrationPointIndex, ThisMethod);
             
-        const double detJ = std::pow(jacobian(0,1),2)*(std::pow(jacobian(1,0),2)+std::pow(jacobian(2,0),2))+std::pow(jacobian(1,1)*jacobian(2,0)-jacobian(1,0)*jacobian(2,1),2)-2*jacobian(0,0)*jacobian(0,1)*(jacobian(1,0)*jacobian(1,1)+jacobian(2,0)*jacobian(2,1))+std::pow(jacobian(0,0),2)*(std::pow(jacobian(1,1),2)+std::pow(jacobian(2,1),2));
+        const double detJ = std::sqrt(std::pow(jacobian(0,1),2.0)*(std::pow(jacobian(1,0),2.0) + std::pow(jacobian(2,0),2.0)) + std::pow(jacobian(1,1)*jacobian(2,0) - jacobian(1,0)*jacobian(2,1),2.0) - 2.0*jacobian(0,0)*jacobian(0,1)*(jacobian(1,0)*jacobian(1,1) + jacobian(2,0)*jacobian(2,1)) + std::pow(jacobian(0,0),2.0)*(std::pow(jacobian(1,1),2.0) + std::pow(jacobian(2,1),2.0)));

```

Comment:
I would separate the sqrt and put a debug check if detJ is negative.....

Reply:
Ok